File: frontend/src/app/domain/usecases/departments/create-department.usecase.ts
import { Injectable, inject } from '@angular/core';
import { Observable, throwError } from 'rxjs';
import { DepartmentRepository } from '@domain/repositories/department.repository';
import { Department } from '@domain/models/department.model';

@Injectable({ providedIn: 'root' })
export class CreateDepartmentUseCase {
  private readonly repo = inject(DepartmentRepository);

  execute(name: string): Observable<Department> {
    const trimmedName = name.trim();
    
    if (!trimmedName) {
      return throwError(() => new Error('Department name cannot be empty'));
    }
    
    return this.repo.create(trimmedName);
  }
}

File: backend/modules/admin/application/departments/create_department_use_case.py
from modules.admin.domain.entities.department import Department
from modules.admin.domain.interfaces.repositories.i_department_repository import (
    IDepartmentRepository,
)
from modules.admin.domain.interfaces.use_cases.departments.i_create_department_use_case import (
    ICreateDepartmentUseCase,
)


class CreateDepartmentUseCase(ICreateDepartmentUseCase):
    def __init__(self, repo: IDepartmentRepository) -> None:
        self._repo = repo

    async def execute(self, name: str) -> Department:
        return await self._repo.create(name)
